File: src/wiki/lint/scanners.ts
// Lint scanner functions — extracted from lint-controller.ts for testability.
// These have no Obsidian API dependencies and can be unit tested directly.

import { parseFrontmatter } from '../../core/frontmatter';
import { getActiveEntityTags, getActiveConceptTags, getActiveSourceTags } from '../../core/tag-vocab';
import { LLMWikiSettings } from '../../types';

export interface ScannerPage {
  path: string;
  content: string;
  basename: string;
}

// Build a set of all known link targets across the vault for dead link detection.
export function buildKnownTargets(allVaultFiles: Array<{ basename: string; path: string }>): { known: Set<string>; knownLower: Set<string> } {
  const known = new Set<string>();
  const knownLower = new Set<string>();
  const addTarget = (t: string) => { known.add(t); knownLower.add(t.toLowerCase()); };
  for (const file of allVaultFiles) {
    const nameWithoutExt = file.basename.replace('.md', '');
    addTarget(file.basename);
    addTarget(nameWithoutExt);
    const relPath = file.path.replace('.md', '');
    addTarget(relPath);
    addTarget(file.path);
    const parts = relPath.split('/');
    for (let i = 1; i < parts.length; i++) {
      const subPath = parts.slice(i).join('/');
      addTarget(subPath);
      addTarget(subPath + '.md');
    }
  }
  return { known, knownLower };
}

// Detect pages with missing aliases (entities & concepts only).
export function detectAliasDeficiency(
  wikiFiles: Array<{ path: string }>,
  pageMap: Map<string, ScannerPage>
): ScannerPage[] {
  const result: ScannerPage[] = [];
  for (const file of wikiFiles) {
    if (file.path.includes('/entities/') || file.path.includes('/concepts/')) {
      const info = pageMap.get(file.path);
      if (info) {
        const fmMatch = info.content.match(/^---\n([\s\S]*?)\n---/);
        if (fmMatch && !fmMatch[1].includes('aliases:')) {
          result.push(info);
        }
      }
    }
  }
  return result;
}

// Scan wiki pages for dead links ([[wikilinks]] pointing to non-existent targets).
// Uses Map<string, {path, content, basename}> which requires no Obsidian types.
export function scanDeadLinks(
  pageMap: Map<string, ScannerPage>,
  knownTargets: Set<string>,
  knownTargetsLower: Set<string>,
  wikiFolder: string
): Array<{ source: string; target: string }> {
  const deadLinks: Array<{ source: string; target: string }> = [];
  // Per-(source,target) dedup so a page referencing the same missing target
  // 4 times shows up as 1 entry, not 4. Diff against (source, target) across
  // pages intentionally stays (different source pages should each list the
  // missing targets they reference, so users can see which sources are affected).
  const seen = new Set<string>();
  const linkRegex = /\[\[([^\]|#]+)(?:[|#][^\]]+)?\]\]/g;
  for (const { path, content } of pageMap.values()) {
    let match: RegExpExecArray | null;
    while ((match = linkRegex.exec(content)) !== null) {
      const target = match[1].trim();
      const targetLower = target.toLowerCase();
      if (!knownTargets.has(target) && !knownTargetsLower.has(targetLower)) {
        // Slug-normalized fallback: "entities/Claude Code" matches "entities/Claude-Code"
        const parts = target.split('/');
        const sluggedBasename = parts[parts.length - 1].replace(/\s+/g, '-');
        const sluggedTarget = [...parts.slice(0, -1), sluggedBasename].join('/');
        const isSlugMatch =
          sluggedTarget !== target &&
          (knownTargets.has(sluggedTarget) || knownTargetsLower.has(sluggedTarget.toLowerCase()));
        if (!isSlugMatch) {
          const source = path.replace(wikiFolder + '/', '').replace('.md', '');
          const key = `${source}::${target}`;
          if (!seen.has(key)) {
            seen.add(key);
            deadLinks.push({ source, target });
          }
        }
      }
    }
    linkRegex.lastIndex = 0;
  }
  return deadLinks;
}

// Detect orphan pages (no incoming links from any wiki page, alias-aware).
export function scanOrphans(
  pageMap: Map<string, ScannerPage>,
  wikiFolder: string
): string[] {
  const incomingLinks = new Map<string, string[]>();
  const linkRegex = /\[\[([^\]|#]+)(?:[|#][^\]]+)?\]\]/g;
  for (const { path, content } of pageMap.values()) {
    const sourceRel = path.replace(wikiFolder + '/', '').replace('.md', '');
    let match: RegExpExecArray | null;
    while ((match = linkRegex.exec(content)) !== null) {
      const target = match[1].trim();
      if (!incomingLinks.has(target)) incomingLinks.set(target, []);
      incomingLinks.get(target)!.push(sourceRel);
    }
    linkRegex.lastIndex = 0;
  }
  const orphans: string[] = [];
  for (const { path, basename, content } of pageMap.values()) {
    const fm = parseFrontmatter(content);
    const aliases = Array.isArray(fm?.aliases) ? fm.aliases : [];
    const relPath = path.replace(wikiFolder + '/', '').replace('.md', '');
    const nameWithoutExt = basename.replace('.md', '');
    const forms = [basename, nameWithoutExt, relPath, ...aliases];
    const parts = relPath.split('/');
    for (let i = 1; i < parts.length; i++) {
      const subPath = parts.slice(i).join('/');
      forms.push(subPath);
      forms.push(subPath + '.md');
    }
    const hasIncoming = forms.some(f => incomingLinks.has(f) || incomingLinks.has(f.toLowerCase()));
    if (!hasIncoming) orphans.push(path);
  }
  return orphans;
}

// ── Quote grounding scanner (Issue #126) ──────────────────────

export interface QuoteGroundingIssue {
  pagePath: string;
  sourcePath?: string;
  quote: string;
  hasSourceLink: boolean;
}

function normalizeQuote(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^\p{L}\p{N}\s]/gu, '')
    .replace(/\s+/g, ' ')
    .trim();
}

function extractMentionsSection(content: string): string | undefined {
  // Match the first "## Mentions in Source" section up to the next ## heading.
  const match = content.match(/##\s+Mentions\s+in\s+Source\s*\n([\s\S]*?)(?=\n##\s|\n*$)/i);
  return match?.[1];
}

function extractSourceBody(content: string): string {
  // Strip YAML frontmatter.
  return content.replace(/^---\n[\s\S]*?\n---\n?/, '');
}

function isQuoteGrounded(quote: string, sourceBody: string): boolean {
  // Tier 1: exact substring.
  if (sourceBody.includes(quote)) return true;
  // Tier 2: normalized substring.
  const normalizedQuote = normalizeQuote(quote);
  if (normalizedQuote.length === 0) return false;
  return normalizeQuote(sourceBody).includes(normalizedQuote);
}

/**
 * Issue #126: programmatic quote-grounding audit. Verifies that every quote
 * listed under a page's `## Mentions in Source` section can be found in the
 * linked (or, for legacy bare quotes, any) source file.
 *
 * Two tolerance tiers:
 *   1. Exact substring match against the source body.
 *   2. Normalized match (case-folded, punctuation stripped, whitespace collapsed).
 *
 * Mentions may have either of these forms:
 *   - "quote text" — [[sources/slug]]     (current format)
 *   - "quote text"                          (legacy format without source link)
 *
 * Legacy bare quotes are accepted if they appear in ANY source file under
 * `wiki/sources/`. This avoids false positives on older pages generated before
 * the source-link suffix was added.
 *
 * Returns a sorted list of ungrounded issues. No file IO, no LLM.
 */
export function scanQuoteGrounding(
  pageMap: Map<string, ScannerPage>,
  sourceMap: Map<string, ScannerPage>,
  wikiFolder: string,
): QuoteGroundingIssue[] {
  const issues: QuoteGroundingIssue[] = [];

  // Pre-build the list of all source bodies for legacy bare-quote fallback.
  const sourceBodies = Array.from(sourceMap.values()).map(s => extractSourceBody(s.content));

  for (const [path, page] of pageMap) {
    if (!path.startsWith(wikiFolder + '/')) continue;

    const mentionsBlock = extractMentionsSection(page.content);
    if (!mentionsBlock) continue;

    // Match lines like: - "quote text" — [[sources/slug]]
    // or legacy:        - "quote text"
    const lineRegex = /^[-*]\s+"([^"]+)"(?:\s*[—-]\s*\[\[([^\]]+)\]\])?\s*$/gm;
    let match: RegExpExecArray | null;
    while ((match = lineRegex.exec(mentionsBlock)) !== null) {
      const quote = match[1].trim();
      const linkTarget = match[2]?.trim();

      if (linkTarget) {
        // Current format: exact source link.
        const sourcePath = linkTarget.endsWith('.md') ? linkTarget : `${linkTarget}.md`;
        const resolvedPath = sourcePath.startsWith(wikiFolder + '/') ? sourcePath : `${wikiFolder}/${sourcePath}`;
        const source = sourceMap.get(resolvedPath);
        const body = source ? extractSourceBody(source.content) : '';
        const grounded = source && isQuoteGrounded(quote, body);
        if (!grounded) {
          issues.push({
            pagePath: path,
            sourcePath: resolvedPath,
            quote,
            hasSourceLink: true,
          });
        }
      } else {
        // Legacy format: accept if quote appears in any source file.
        const grounded = sourceBodies.some(body => isQuoteGrounded(quote, body));
        if (!grounded) {
          issues.push({
            pagePath: path,
            quote,
            hasSourceLink: false,
          });
        }
      }
    }
  }

  issues.sort((a, b) => {
    const pathCmp = a.pagePath.localeCompare(b.pagePath);
    if (pathCmp !== 0) return pathCmp;
    return a.quote.localeCompare(b.quote);
  });

  return issues;
}

// ── Tag vocabulary violation scanner (Issue #85 v7) ───────────

export type TagViolationPageType = 'entity' | 'concept' | 'source';

export interface TagViolation {
  path: string;
  pageType: TagViolationPageType;
  title: string;
  currentTags: string[];
  invalidTags: string[];   // subset of currentTags that are NOT in the active vocabulary
}

/**
 * Issue #85 v7: programmatic tag-vocabulary audit. Pure function. Walks
 * the same pageMap used by the other Lint scanners and reports every
 * entity / concept / source page whose frontmatter `tags` array
 * contains at least one value that is not in the active vocabulary.
 *
 * Active vocabulary is resolved via the existing getActive*Tags
 * helpers so this scanner automatically tracks Issue #85 v6 settings
 * (default vs custom mode, plus the new static source-page taxonomy
 * VALID_SOURCE_TAGS).
 *
 * Returns an empty array when everything is clean. No file IO, no LLM.
 * Sort: by path, so the Lint report is deterministic.
 */
export function scanTagViolations(
  pageMap: Map<string, ScannerPage>,
  settings: LLMWikiSettings,
): TagViolation[] {
  const validEntity = new Set(getActiveEntityTags(settings));
  const validConcept = new Set(getActiveConceptTags(settings));
  const validSource = new Set(getActiveSourceTags(settings));
  const violations: TagViolation[] = [];

  for (const [path, page] of pageMap) {
    const fm = parseFrontmatter(page.content);
    if (!fm) continue;
    const pageType = fm.type as TagViolationPageType | 'comparison' | 'overview' | undefined;
    if (pageType !== 'entity' && pageType !== 'concept' && pageType !== 'source') continue;

    const validSet =
      pageType === 'entity' ? validEntity :
      pageType === 'concept' ? validConcept :
      validSource;

    // tags can be a string (YAML scalar) or array. parseFrontmatter
    // returns a string for scalar and string[] for array. Accept both.
    // (The runtime type is broad; tsc narrows it to never here, so we
    // explicitly cast to unknown then back to the union we actually
    // handle below.)
    const rawTags: unknown = (fm as Record<string, unknown>).tags;
    let currentTags: string[];
    if (Array.isArray(rawTags)) {
      currentTags = rawTags.map(t => String(t).trim()).filter(t => t.length > 0);
    } else if (typeof rawTags === 'string' && rawTags.length > 0) {
      currentTags = [rawTags.trim()];
    } else {
      continue; // empty / no tags → not a violation
    }

    const invalidTags = currentTags.filter(t => !validSet.has(t));
    if (invalidTags.length > 0) {
      violations.push({
        path,
        pageType,
        title: typeof fm.title === 'string' ? fm.title : page.basename.replace(/\.md$/, ''),
        currentTags,
        invalidTags,
      });
    }
  }

  violations.sort((a, b) => a.path.localeCompare(b.path));
  return violations;
}

// ── Hub link density scanner (Issue #157 / #175, v1.23.0 P1-6) ──────────
// Re-exported from core/ because the scanner needs a Graph (wiki-link
// structure), which the other scanners don't. Keeping the
// implementation in core/ preserves the pure-function convention and
// makes the algorithm unit-testable without an Obsidian dependency.

export { scanHubLinkDensity, type HubLinkDensityIssue, type HubLinkDensityOptions } from '../../core/hub-link-distinctiveness';

File: python_lib/llm_wiki/wiki/lint/scanners.py
"""
Lint scanners — pure, no-LLM, no-IO functions.
Port of src/wiki/lint/scanners.ts and src/wiki/lint/utils.ts
"""
from __future__ import annotations

import re
from typing import Dict, List, Optional, Set, Tuple

from ...core.frontmatter import parse_frontmatter
from ...types import LLMWikiSettings, VALID_ENTITY_TAGS, VALID_CONCEPT_TAGS, VALID_SOURCE_TAGS


# ── Types ─────────────────────────────────────────────────────────────────────

class ScannerPage:
    def __init__(self, path: str, content: str, basename: str) -> None:
        self.path = path
        self.content = content
        self.basename = basename


# ── Utils ─────────────────────────────────────────────────────────────────────

EMPTY_CONTENT_STRIP = re.compile(r'[#*\-_>\s\n\[\]|—]')
MIN_SUBSTANTIVE_CHARS = 50
STUB_MARKER = "Stub created by Fix Dead Links"


def is_page_empty(content: str) -> bool:
    if STUB_MARKER in content:
        return True
    text_body = re.sub(r"---[\s\S]*?---", "", content)
    text_body = EMPTY_CONTENT_STRIP.sub("", text_body).strip()
    return len(text_body) < MIN_SUBSTANTIVE_CHARS


def fix_double_nested_wiki_links(content: str) -> Tuple[int, str]:
    """Collapse [[[[target|display]]]] → [[target|display]]."""
    fixed = 0

    def replacer(m: re.Match) -> str:
        nonlocal fixed
        fixed += 1
        target = m.group(1)
        display = m.group(2)
        return f"[[{target}|{display}]]" if display else f"[[{target}]]"

    result = re.sub(
        r"\[\[\[\[([^\]|#]+)(?:\|([^\]]+))?\]\]\]\]",
        replacer,
        content,
    )
    return fixed, result


def detect_polluted_pages(
    pages: List[Dict],
) -> List[Dict]:
    """Detect pages whose title/basename starts with a folder prefix."""
    polluted = []
    for p in pages:
        title = p.get("title", "")
        m = re.match(r"^(entities|concepts|sources)([^\s\-_a-zA-Z0-9])", title)
        if m:
            clean_title = re.sub(r"^(entities|concepts|sources)", "", title)
            polluted.append({"path": p["path"], "title": title, "cleanTitle": clean_title})
    return polluted


def normalize_frontmatter_dates(content: str, date_str: str) -> str:
    fm_match = re.match(r"^---\n([\s\S]*?)\n---", content)
    if not fm_match:
        return content
    fm_content = fm_match.group(1)
    if re.search(r"^updated:\s*.+$", fm_content, re.MULTILINE):
        new_fm = re.sub(r"^updated:\s*.+$", f"updated: {date_str}", fm_content, flags=re.MULTILINE)
    else:
        new_fm = fm_content + f"\nupdated: {date_str}"
    return re.sub(r"^---\n[\s\S]*?\n---", f"---\n{new_fm}\n---", content, count=1)


def build_section_labels_hint(settings: LLMWikiSettings) -> str:
    from ...wiki.prompts import get_section_labels
    labels = get_section_labels(getattr(settings, "wiki_language", "en"))
    entity_labels = "\n".join([
        f"- {labels['basic_information']}",
        f"- {labels['description']}",
        f"- {labels['related_entities']}",
        f"- {labels['related_concepts']}",
        f"- {labels['mentions_in_source']}",
    ])
    concept_labels = "\n".join([
        f"- {labels['definition']}",
        f"- {labels['key_characteristics']}",
        f"- {labels['applications']}",
        f"- {labels['related_concepts']}",
        f"- {labels['related_entities']}",
        f"- {labels['mentions_in_source']}",
    ])
    source_labels = "\n".join([
        f"- {labels['source']}",
        f"- {labels['core_content']}",
        f"- {labels['key_entities']}",
        f"- {labels['key_concepts']}",
        f"- {labels['main_points']}",
    ])
    return (
        f"Entity pages use:\n{entity_labels}\n\n"
        f"Concept pages use:\n{concept_labels}\n\n"
        f"Source pages use:\n{source_labels}"
    )


def escape_regex(s: str) -> str:
    return re.escape(s)


# ── Known-targets set ─────────────────────────────────────────────────────────

def build_known_targets(
    all_files: List[Dict],  # [{"basename": str, "path": str}]
) -> Tuple[Set[str], Set[str]]:
    known: Set[str] = set()
    known_lower: Set[str] = set()

    def add(t: str) -> None:
        known.add(t)
        known_lower.add(t.lower())

    for f in all_files:
        basename = f.get("basename", "")
        path = f.get("path", "")
        name_no_ext = basename.replace(".md", "")
        add(basename)
        add(name_no_ext)
        rel_path = path.replace(".md", "")
        add(rel_path)
        add(path)
        parts = rel_path.split("/")
        for i in range(1, len(parts)):
            sub = "/".join(parts[i:])
            add(sub)
            add(sub + ".md")

    return known, known_lower


# ── Dead-link scanner ─────────────────────────────────────────────────────────

LINK_RE = re.compile(r"\[\[([^\]|#]+)(?:[|#][^\]]+)?\]\]")


def scan_dead_links(
    page_map: Dict[str, ScannerPage],
    known_targets: Set[str],
    known_targets_lower: Set[str],
    wiki_folder: str,
) -> List[Dict]:
    """Return list of {source, target} for every dead [[wikilink]]."""
    dead: List[Dict] = []
    seen: Set[str] = set()

    for page in page_map.values():
        for m in LINK_RE.finditer(page.content):
            target = m.group(1).strip()
            target_lower = target.lower()
            if target in known_targets or target_lower in known_targets_lower:
                continue
            # Slug-normalised fallback: spaces → hyphens
            parts = target.split("/")
            slugged_basename = parts[-1].replace(" ", "-")
            slugged = "/".join(parts[:-1] + [slugged_basename])
            if (
                slugged != target
                and (slugged in known_targets or slugged.lower() in known_targets_lower)
            ):
                continue
            source = page.path.replace(wiki_folder + "/", "").replace(".md", "")
            key = f"{source}::{target}"
            if key not in seen:
                seen.add(key)
                dead.append({"source": source, "target": target})

    return dead


# ── Orphan scanner ────────────────────────────────────────────────────────────

def scan_orphans(
    page_map: Dict[str, ScannerPage],
    wiki_folder: str,
) -> List[str]:
    """Return paths of pages with no incoming wiki-links (alias-aware)."""
    incoming: Dict[str, List[str]] = {}

    for page in page_map.values():
        source_rel = page.path.replace(wiki_folder + "/", "").replace(".md", "")
        for m in LINK_RE.finditer(page.content):
            target = m.group(1).strip()
            incoming.setdefault(target, []).append(source_rel)

    orphans: List[str] = []
    for page in page_map.values():
        fm = parse_frontmatter(page.content) or {}
        aliases = fm.get("aliases", []) if isinstance(fm.get("aliases"), list) else []
        rel_path = page.path.replace(wiki_folder + "/", "").replace(".md", "")
        name_no_ext = page.basename.replace(".md", "")
        forms = [page.basename, name_no_ext, rel_path] + list(aliases)
        parts = rel_path.split("/")
        for i in range(1, len(parts)):
            sub = "/".join(parts[i:])
            forms.append(sub)
            forms.append(sub + ".md")

        has_incoming = any(
            f in incoming or f.lower() in {k.lower() for k in incoming}
            for f in forms
        )
        if not has_incoming:
            orphans.append(page.path)

    return orphans


# ── Alias-deficiency scanner ──────────────────────────────────────────────────

def detect_alias_deficiency(
    wiki_files: List[Dict],
    page_map: Dict[str, ScannerPage],
) -> List[ScannerPage]:
    """Return entity/concept pages that have no aliases in their frontmatter."""
    result: List[ScannerPage] = []
    for f in wiki_files:
        path = f.get("path", "")
        if "/entities/" in path or "/concepts/" in path:
            info = page_map.get(path)
            if info:
                fm_match = re.match(r"^---\n([\s\S]*?)\n---", info.content)
                if fm_match and "aliases:" not in fm_match.group(1):
                    result.append(info)
    return result


# ── Quote grounding scanner ───────────────────────────────────────────────────

class QuoteGroundingIssue:
    def __init__(
        self,
        page_path: str,
        quote: str,
        has_source_link: bool,
        source_path: Optional[str] = None,
    ) -> None:
        self.page_path = page_path
        self.quote = quote
        self.has_source_link = has_source_link
        self.source_path = source_path


def _normalize_quote(text: str) -> str:
    text = text.lower()
    text = re.sub(r"[^\w\s]", "", text, flags=re.UNICODE)
    return re.sub(r"\s+", " ", text).strip()


def _extract_mentions_section(content: str) -> Optional[str]:
    m = re.search(
        r"##\s+Mentions\s+in\s+Source\s*\n([\s\S]*?)(?=\n##\s|\n*$)",
        content,
        re.IGNORECASE,
    )
    return m.group(1) if m else None


def _extract_source_body(content: str) -> str:
    return re.sub(r"^---\n[\s\S]*?\n---\n?", "", content)


def _is_quote_grounded(quote: str, source_body: str) -> bool:
    if quote in source_body:
        return True
    return _normalize_quote(quote) in _normalize_quote(source_body)


def scan_quote_grounding(
    page_map: Dict[str, ScannerPage],
    source_map: Dict[str, ScannerPage],
    wiki_folder: str,
) -> List[QuoteGroundingIssue]:
    issues: List[QuoteGroundingIssue] = []
    source_bodies = [_extract_source_body(s.content) for s in source_map.values()]

    line_re = re.compile(
        r'^[-*]\s+"([^"]+)"(?:\s*[—\-]\s*\[\[([^\]]+)\]\])?\s*$',
        re.MULTILINE,
    )

    for path, page in page_map.items():
        if not path.startswith(wiki_folder + "/"):
            continue
        mentions_block = _extract_mentions_section(page.content)
        if not mentions_block:
            continue

        for m in line_re.finditer(mentions_block):
            quote = m.group(1).strip()
            link_target = m.group(2).strip() if m.group(2) else None

            if link_target:
                src_path = link_target if link_target.endswith(".md") else link_target + ".md"
                resolved = src_path if src_path.startswith(wiki_folder + "/") else f"{wiki_folder}/{src_path}"
                source = source_map.get(resolved)
                body = _extract_source_body(source.content) if source else ""
                if not (source and _is_quote_grounded(quote, body)):
                    issues.append(QuoteGroundingIssue(path, quote, True, resolved))
            else:
                grounded = any(_is_quote_grounded(quote, b) for b in source_bodies)
                if not grounded:
                    issues.append(QuoteGroundingIssue(path, quote, False))

    issues.sort(key=lambda i: (i.page_path, i.quote))
    return issues


# ── Tag-violation scanner ─────────────────────────────────────────────────────

class TagViolation:
    def __init__(
        self,
        path: str,
        page_type: str,
        title: str,
        current_tags: List[str],
        invalid_tags: List[str],
    ) -> None:
        self.path = path
        self.page_type = page_type
        self.title = title
        self.current_tags = current_tags
        self.invalid_tags = invalid_tags


def _get_active_entity_tags(settings: LLMWikiSettings) -> List[str]:
    mode = getattr(settings, "tag_vocabulary_mode", "default")
    if mode == "custom":
        raw = getattr(settings, "custom_entity_tags", "") or ""
        tags = [t.strip() for t in raw.split(",") if t.strip()]
        return tags if tags else list(VALID_ENTITY_TAGS)
    return list(VALID_ENTITY_TAGS)


def _get_active_concept_tags(settings: LLMWikiSettings) -> List[str]:
    mode = getattr(settings, "tag_vocabulary_mode", "default")
    if mode == "custom":
        raw = getattr(settings, "custom_concept_tags", "") or ""
        tags = [t.strip() for t in raw.split(",") if t.strip()]
        return tags if tags else list(VALID_CONCEPT_TAGS)
    return list(VALID_CONCEPT_TAGS)


def _get_active_source_tags(settings: LLMWikiSettings) -> List[str]:
    return list(VALID_SOURCE_TAGS)


def scan_tag_violations(
    page_map: Dict[str, ScannerPage],
    settings: LLMWikiSettings,
) -> List[TagViolation]:
    valid_entity = set(_get_active_entity_tags(settings))
    valid_concept = set(_get_active_concept_tags(settings))
    valid_source = set(_get_active_source_tags(settings))

    violations: List[TagViolation] = []
    for path, page in page_map.items():
        fm = parse_frontmatter(page.content)
        if not fm:
            continue
        page_type = fm.get("type", "")
        if page_type not in ("entity", "concept", "source"):
            continue

        valid_set = (
            valid_entity if page_type == "entity"
            else valid_concept if page_type == "concept"
            else valid_source
        )

        raw_tags = fm.get("tags", [])
        if isinstance(raw_tags, list):
            current_tags = [str(t).strip() for t in raw_tags if str(t).strip()]
        elif isinstance(raw_tags, str) and raw_tags.strip():
            current_tags = [raw_tags.strip()]
        else:
            continue

        invalid = [t for t in current_tags if t not in valid_set]
        if invalid:
            title = page.basename.replace(".md", "")
            violations.append(TagViolation(path, page_type, title, current_tags, invalid))

    violations.sort(key=lambda v: v.path)
    return violations
